src/vos/vos_pool_scrub.c

```diff
@@ -0,0 +1,735 @@
+	data.iov_len = data_len;
+
+	/* Fetch data */
+	iter = vos_hdl2iter(ih);
+	oiter = vos_iter2oiter(iter);
+	bio_ctx = oiter->it_obj->obj_cont->vc_pool->vp_io_ctxt;
+	rc = bio_read(bio_ctx, biov->bi_addr, &data);
+
+	/* if bio_read of NVME then it might have yielded */
+	if (bio_iov2media(biov) == DAOS_MEDIA_NVME)
+		ctx->sc_did_yield = true;
+
+	if (rc == -DER_CSUM) {
+		/* Already know this is corrupt so just return */
+		D_GOTO(out, rc = DER_SUCCESS);
+	} else if (rc != 0) {
+		D_WARN("Unable to fetch data for scrubber");
+		D_GOTO(out, rc);
+	}
+
+	rc = iod->iod_type == DAOS_IOD_ARRAY ?
+	     sc_verify_recx(ctx, &data) :
+	     sc_verify_sv(ctx, &data);
+
+	if (rc == -DER_CSUM) {
+		D_WARN("Checksum scrubber found corruption");
+		sc_raise_ras(ctx);
+		sc_m_pool_corr_inc(ctx);
+		rc = sc_mark_corrupt(ctx);
+	}
+
+out:
+	D_FREE(data.iov_buf);
+
+	return rc;
+}
+
+static void
+sc_obj_val_setup(struct scrub_ctx *ctx, vos_iter_entry_t *entry,
+		 vos_iter_type_t type, vos_iter_param_t *param,
+		 daos_handle_t ih)
+{
+	ctx->sc_cur_oid = param->ip_oid;
+	ctx->sc_dkey = param->ip_dkey;
+	ctx->sc_epoch = entry->ie_epoch;
+	ctx->sc_minor_epoch = entry->ie_minor_epc;
+
+	ctx->sc_iod.iod_size = entry->ie_rsize;
+	ctx->sc_iod.iod_nr = 1;
+	ctx->sc_iod.iod_type = type == VOS_ITER_RECX ? DAOS_IOD_ARRAY :
+			       DAOS_IOD_SINGLE;
+	ctx->sc_iod.iod_name = param->ip_akey;
+	ctx->sc_iod.iod_recxs = &entry->ie_recx;
+
+	ctx->sc_csum_to_verify = &entry->ie_csum;
+
+	ctx->sc_vos_iter_handle = ih;
+}
+
+static bool
+oids_are_same(daos_unit_oid_t a, daos_unit_oid_t b)
+{
+	return daos_unit_oid_compare(a, b) == 0;
+}
+
+static bool
+keys_are_same(daos_key_t key1, daos_key_t key2)
+{
+	if (key1.iov_len != key2.iov_len)
+		return 0;
+
+	return memcmp(key1.iov_buf, key2.iov_buf, key1.iov_len) == 0;
+}
+
+static bool
+epoch_eq(daos_epoch_t a, daos_epoch_t b)
+{
+	return a == b;
+}
+
+static bool
+recx_eq(const daos_recx_t *a, const daos_recx_t *b)
+{
+	if (a == NULL || b == NULL)
+		return false;
+
+	return a->rx_nr == b->rx_nr && a->rx_idx == b->rx_idx;
+}
+
+static bool
+sc_value_has_been_seen(struct scrub_ctx *ctx, vos_iter_entry_t *entry,
+		       vos_iter_type_t type)
+{
+	if (type == VOS_ITER_RECX &&
+	    !recx_eq(ctx->sc_iod.iod_recxs, &entry->ie_recx))
+		return false;
+	return epoch_eq(ctx->sc_epoch, entry->ie_epoch) &&
+	       epoch_eq(ctx->sc_minor_epoch, entry->ie_minor_epc);
+}
+
+static void
+sc_obj_value_reset(struct  scrub_ctx *ctx)
+{
+	ctx->sc_epoch = 0;
+	ctx->sc_minor_epoch = 0;
+}
+
+/** vos_iter_cb_t */
+static int
+obj_iter_scrub_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		  vos_iter_type_t type, vos_iter_param_t *param,
+		  void *cb_arg, unsigned int *acts)
+{
+	struct scrub_ctx	*ctx = cb_arg;
+	int			 rc;
+
+	if (sc_cont_is_stopping(ctx)) {
+		C_TRACE("Container is stopping.");
+		return 1;
+	}
+
+	if (ctx->sc_pool->sp_scrub_sched == DAOS_SCRUB_SCHED_OFF) {
+		C_TRACE("scrubbing is off now");
+		return 1;
```

Comment:
You'd use different return value here, for container stopping(destroy), only the container scrubbing should be aborted, for turn off scrubbing, the whole scrubbing should be aborted.